File: android/app/src/main/kotlin/com/example/hotel_reception/procesor/model/World.kt
package com.example.hotel_reception.procesor.model

import android.os.Handler
import android.util.Log
import com.example.hotel_reception.procesor.WorldsClock
import com.example.hotel_reception.procesor.dto.WorldSpeed
import java.util.*
import java.util.concurrent.TimeUnit

class World : WorldsClock {
    private val calendar: Calendar = Calendar.getInstance()
    private val listeners: MutableList<WorldDateChanged> = mutableListOf()
    private var timer: Timer? = null
    private var worldSpeed = WorldSpeed.normal
    private fun nextDay() {
        calendar.add(Calendar.DATE, 1)
        notifyDateChanged()
    }

    override fun getDate(): Date = calendar.time

    override fun getDayOfMonth(): Int = calendar.get(Calendar.DAY_OF_MONTH)

    private fun notifyDateChanged() {
        listeners.forEach(WorldDateChanged::onDateChanged)
    }

    fun subscribeOnDateChanges(listener: WorldDateChanged) {
        listeners.add(listener)
    }

    override fun getSeason(): Season {
        return when (calendar.get(Calendar.DAY_OF_YEAR)) {
            in 1..90 -> Season.winter
            in 90..180 -> Season.spring
            in 180..270 -> Season.summer
            else -> Season.autmn
        }
    }

    fun getCurrentMultiply(): Double = getSeason().priceMultiply

    fun startWorld() {
        Log.i(logTag, "currentSpeed = ${worldSpeed.time}")
        timer = Timer()
        timer?.schedule(object : TimerTask() {
            override fun run() {
                nextDay()
            }

        }, 0, TimeUnit.MILLISECONDS.toMillis(worldSpeed.time))

    }

    override fun getCurrentSpeed() = worldSpeed
    override fun nextSpeed() {
        Log.i(logTag, "nextSpeed")
        worldSpeed = WorldSpeed.values().let {
            it.indexOf(worldSpeed).let { index ->
                if (it.indexOf(worldSpeed) == it.lastIndex) {
                    it.first()
                } else {
                    it[index+1]
                }
            }
        }
        timer?.cancel()
        startWorld()
    }


}

//inline fun Timer.schedule(
//    delay: Long,
//    crossinline action: TimerTask.() -> Unit
//): TimerTask {
//}

enum class Season(val priceMultiply: Double) {
    winter(1.5),
    spring(1.0),
    summer(1.5),
    autmn(1.0),
}

interface WorldDateChanged {
    fun onDateChanged()
}

/// [date, season,  dayOfWeek]

File: ios/Runner/Processor/model/world/World.swift
//
//  World.swift
//  Runner
//
//  Created by Serhii Rakovskyi on 17.08.2022.
//

import Foundation

class World : WorldsClock {
    private var date = Date()
    private var listeners: Array<WorldChangedListener> = []
    private var speed = WorldSpeed.normal
    private var timer: Timer?
    
    private func nextDay() {
        var dayComponent = DateComponents()
        dayComponent.day = 1
        date = Calendar.current.date(byAdding: dayComponent , to: date)!
        notifyListeners()
    }
    
    private func notifyListeners() {
        print("date changed")
        listeners.forEach {
            $0.onDateChanged()
        }
    }
    
    
    func getDate() -> Date {
        return date;
    }
    
    func getDayOfMonth() -> Int {
        return Calendar.current.dateComponents([.day], from: date).day!
    }
    
    func getCurrentSpeed() -> WorldSpeed {
        return speed;
    }
    
    func getSeason() -> Season {
        let month = Calendar.current.dateComponents([.month], from: date).month!
        switch month {
        case  0...3 : return Season.winter(1.5)
        case  4...6 : return Season.spring(1)
        case  6...9 : return Season.summer(1.5)
        case  9...11 : return Season.autmn(1)
        default:
            return Season.winter(1.5)
        }
    }
    
    func getMultiply() -> Double {
        switch getSeason() {
        case let .autmn(multiply): return multiply
        case let .winter(multiply): return multiply
        case let .summer(multiply): return multiply
        case let .spring(multiply): return multiply
        }
    }
    func startWorld() {
        if #available(iOS 10.0, *) {
            print("startWorld")
            timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { t in
                self.nextDay()
            }
        } else {
            // Fallback on earlier versions
            
        }
    }
    
    func nextSpeed() {
        let values = [WorldSpeed.normal, WorldSpeed.fast, WorldSpeed.faster];
        let index = values.firstIndex(of: speed)!
        if values.endIndex == index
        {speed = values.first!}
        else
        {speed = values[index+1]}
        timer?.invalidate()
        startWorld()
    }
    
    func subscribeOnDateChanges(listener: WorldChangedListener) {
        listeners.append(listener)
    }
}
